<a href="https://colab.research.google.com/github/Jandsy/ml_finance_imperial/blob/main/Coursework/CourseWork.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# **<center>Machine Learning and Finance </center>**


## <center> CourseWork 2024 - StatArb </center>


# Libraries

In [8]:
import requests 
from bs4 import BeautifulSoup

import yfinance as yf
import pandas as pd


In this coursework, you will delve into and replicate selected elements of the research detailed in the paper **[End-to-End Policy Learning of a Statistical Arbitrage Autoencoder Architecture](https://arxiv.org/pdf/2402.08233.pdf)**. **However, we will not reproduce the entire study.**

## Overview

This study redefines Statistical Arbitrage (StatArb) by combining Autoencoder architectures and policy learning to generate trading strategies. Traditionally, StatArb involves finding the mean of a synthetic asset through classical or PCA-based methods before developing a mean reversion strategy. However, this paper proposes a data-driven approach using an Autoencoder trained on US stock returns, integrated into a neural network representing portfolio trading policies to output portfolio allocations directly.


## Coursework Goal

This coursework will replicate these results, providing hands-on experience in implementing and evaluating this innovative end-to-end policy learning Autoencoder within financial trading strategies.

## Outline

- [Data Preparation and Exploration](#Data-Preparation-and-Exploration)
- [Fama French Analysis](#Fama-French-Analysis)
- [PCA Analysis](#PCA-Analysis)
- [Ornstein Uhlenbeck](#Ornstein-Uhlenbeck)
- [Autoencoder Analysis](#Autoencoder-Analysis)



**Description:**
The Coursework is graded on a 100 point scale and is divided into five  parts. Below is the mark distribution for each question:

| **Problem**  | **Question**          | **Number of Marks** |
|--------------|-----------------------|---------------------|
| **Part A**   | Question 1            | 4                   |
|              | Question 2            | 1                   |
|              | Question 3            | 3                   |
|              | Question 4            | 3                   |
|              | Question 5            | 1                   |
|              | Question 6            | 3                   |
|**Part  B**    | Question 7           | 1                   |
|              | Question 8            | 5                   |
|              | Question 9            | 4                   |
|              | Question 10           | 5                   |
|              | Question 11           | 2                   |
|              | Question 12           | 3                   |
|**Part  C**    | Question 13          | 3                   |
|              | Question 14           | 1                   |
|              | Question 15           | 3                   |
|              | Question 16           | 2                   |
|              | Question 17           | 7                   |
|              | Question 18           | 6                   |
|              | Question 19           | 3                   |
|  **Part  D** | Question 20           | 3                   |
|              | Question 21           | 5                   |
|              | Question 22           | 2                   |
|  **Part  E** | Question 23           | 2                   |
|              | Question 24           | 1                   |
|              | Question 25           | 3                   |
|              | Question 26           | 10                  |
|              | Question 27           | 1                   |
|              | Question 28           | 3                   |
|              | Question 29           | 3                   |
|              | Question 30           | 7                   |




Please read the questions carefully and do your best. Good luck!

## Objectives



## 1. Data Preparation and Exploration
Collect, clean, and prepare US stock return data for analysis.

## 2. Fama French Analysis
Utilize Fama French Factors to isolate the idiosyncratic components of stock returns, differentiating them from market-wide effects. This analysis helps in understanding the unique characteristics of individual stocks relative to broader market trends.

## 3. PCA Analysis
Employ Principal Component Analysis (PCA) to identify hidden structures and reduce dimensionality in the data. This method helps in extracting significant patterns that might be obscured in high-dimensional datasets.

## 4. Ornstein-Uhlenbeck Process
Analyze mean-reverting behavior in stock prices using the Ornstein-Uhlenbeck process. This stochastic process is useful for modeling and forecasting based on the assumption that prices will revert to a long-term mean.

## 5. Building a Basic Autoencoder Model
Construct and train a standard Autoencoder to extract residual idiosyncratic risk.








# Data Preparation and Exploration


---
<font color=green>Q1: (4 Marks)</font>
<br><font color='green'>
Write a Python function that accepts a URL parameter and retrieves the NASDAQ-100 companies and their ticker symbols by scraping the relevant Wikipedia page using **[Requests](https://pypi.org/project/requests/)** and **[BeautifulSoup](https://www.crummy.com/software/BeautifulSoup/bs4/doc/)**. Your function should return the data as a list of tuples, with each tuple containing the company name and its ticker symbol. Then, call your function with the appropriate Wikipedia page URL and print the data in a 'Company: Ticker' format.

</font>

---


In [9]:
def get_nasdaq_100(url):
    response = requests.get(url)
    soup = BeautifulSoup(response.text, 'html.parser')
    table = soup.find('table', {'class': 'wikitable sortable'})
    rows = table.find_all('tr')[1:]  # Skip the header row

    companies = []
    for row in rows:
        cols = row.find_all('td')
        company = cols[0].text.strip()
        ticker = cols[1].text.strip()
        companies.append((company, ticker))

    return companies

url = 'https://en.wikipedia.org/wiki/NASDAQ-100'
companies = get_nasdaq_100(url)

# Print the data in a 'Company: Ticker' format
for company, ticker in companies:
    print(f'Company: {company}, Ticker: {ticker}')

Company: Adobe Inc., Ticker: ADBE
Company: ADP, Ticker: ADP
Company: Airbnb, Ticker: ABNB
Company: Alphabet Inc. (Class A), Ticker: GOOGL
Company: Alphabet Inc. (Class C), Ticker: GOOG
Company: Amazon, Ticker: AMZN
Company: Advanced Micro Devices Inc., Ticker: AMD
Company: American Electric Power, Ticker: AEP
Company: Amgen, Ticker: AMGN
Company: Analog Devices, Ticker: ADI
Company: Ansys, Ticker: ANSS
Company: Apple Inc., Ticker: AAPL
Company: Applied Materials, Ticker: AMAT
Company: ASML Holding, Ticker: ASML
Company: AstraZeneca, Ticker: AZN
Company: Atlassian, Ticker: TEAM
Company: Autodesk, Ticker: ADSK
Company: Baker Hughes, Ticker: BKR
Company: Biogen, Ticker: BIIB
Company: Booking Holdings, Ticker: BKNG
Company: Broadcom Inc., Ticker: AVGO
Company: Cadence Design Systems, Ticker: CDNS
Company: CDW Corporation, Ticker: CDW
Company: Charter Communications, Ticker: CHTR
Company: Cintas, Ticker: CTAS
Company: Cisco, Ticker: CSCO
Company: Coca-Cola Europacific Partners, Ticker: CCEP

---
Q2: (1 Mark)

Given a list of tuples representing NASDAQ-100 companies (where each tuple contains a company name and its ticker symbol), write a Python script to extract all ticker symbols into a separate list called `tickers_list`.

---


In [10]:
tickers_list = [ticker for company, ticker in companies]

---
Q3: (3 Marks)

Using **[yfinance](https://pypi.org/project/yfinance/)** library, write a Python script that accepts a list of stock ticker symbols. For each symbol, download the adjusted closing price data, store it in a dictionary with the ticker symbol as the key, and then convert the final dictionary into a Pandas DataFrame. Handle any errors encountered during data retrieval by printing a message indicating which symbol failed

---

In [23]:
def get_adjusted_close_prices(tickers):
    data_dict = {}
    for ticker in tickers:
        try:
            data = yf.download(ticker, start='2000-01-01', end='2024-05-21')
            data_dict[ticker] = data['Adj Close']
        except Exception as e:
            print(f"Failed to download data for {ticker}. Reason: {e}")

    df = pd.DataFrame(data_dict)
    return df

df = get_adjusted_close_prices(tickers_list)

[*********************100%%**********************]  1 of 1 completed
[*********************100%%**********************]  1 of 1 completed
[*********************100%%**********************]  1 of 1 completed
[*********************100%%**********************]  1 of 1 completed
[*********************100%%**********************]  1 of 1 completed
[*********************100%%**********************]  1 of 1 completed
[*********************100%%**********************]  1 of 1 completed
[*********************100%%**********************]  1 of 1 completed
[*********************100%%**********************]  1 of 1 completed
[*********************100%%**********************]  1 of 1 completed
[*********************100%%**********************]  1 of 1 completed
[*********************100%%**********************]  1 of 1 completed
[*********************100%%**********************]  1 of 1 completed
[*********************100%%**********************]  1 of 1 completed
[*********************100%%*******

In [24]:
# save the data to a csv file
df.to_csv('nasdaq_100_data.csv')
df

ADBE         ADP        ABNB       GOOGL        GOOG  \
Date                                                                     
2000-01-03   16.274672   24.781528         NaN         NaN         NaN   
2000-01-04   14.909400   24.781528         NaN         NaN         NaN   
2000-01-05   15.204174   24.543242         NaN         NaN         NaN   
2000-01-06   15.328286   24.870886         NaN         NaN         NaN   
2000-01-07   16.072985   25.436808         NaN         NaN         NaN   
...                ...         ...         ...         ...         ...   
2024-05-14  475.950012  245.500000  146.699997  170.339996  171.929993   
2024-05-15  485.350006  246.619995  145.800003  172.509995  173.880005   
2024-05-16  482.880005  250.059998  147.190002  174.179993  175.429993   
2024-05-17  483.429993  252.330002  145.660004  176.059998  177.289993   
2024-05-20  484.690002  251.779999  146.369995  176.919998  178.460007   

                  AMZN         AMD        AEP        AMGN         ADI  ...  \
Date                                                                   ...   
2000-01-03    4.468750   15.500000  10.773263   44.925060   28.438284  ...   
2000-01-04    4.096875   14.625000  10.901772   41.489872   26.999632  ...   
2000-01-05    3.487500   15.000000  11.308716   42.917492   27.393780  ...   
2000-01-06    3.278125   16.000000  11.372970   43.631294   26.644884  ...   
2000-01-07    3.478125   16.250000  11.522890   48.538692   27.393780  ...   
...                ...         ...        ...         ...         ...  ...   
2024-05-14  187.070007  153.160004  90.790001  309.213806  211.940002  ...   
2024-05-15  185.990005  159.669998  91.970001  316.790009  215.750000  ...   
2024-05-16  183.630005  162.619995  92.540001  314.720001  214.119995  ...   
2024-05-17  184.699997  164.470001  92.669998  312.470001  214.080002  ...   
2024-05-20  183.539993  166.330002  92.589996  314.540009  217.479996  ...   

                  TSLA         TXN        TTD        VRSK        VRTX  \
Date                                                                    
2000-01-03         NaN   32.968735        NaN         NaN   18.781250   
2000-01-04         NaN   31.566660        NaN         NaN   17.281250   
2000-01-05         NaN   30.805527        NaN         NaN   17.000000   
2000-01-06         NaN   29.964285        NaN         NaN   16.750000   
2000-01-07         NaN   30.124537        NaN         NaN   18.218750   
...                ...         ...        ...         ...         ...   
2024-05-14  177.550003  191.130005  86.180000  246.929993  428.589996   
2024-05-15  173.990005  195.529999  90.250000  247.839996  437.489990   
2024-05-16  174.839996  194.970001  93.190002  251.479996  440.640015   
2024-05-17  177.460007  195.020004  94.779999  251.619995  445.209991   
2024-05-20  174.949997  199.199997  97.500000  252.410004  445.869995   

                  WBA   WBD        WDAY        XEL          ZS  
Date                                                            
2000-01-03  17.076204   NaN         NaN   6.977996         NaN  
2000-01-04  16.440989   NaN         NaN   7.138672         NaN  
2000-01-05  16.627823   NaN         NaN   7.414119         NaN  
2000-01-06  16.142059   NaN         NaN   7.345260         NaN  
2000-01-07  16.553080   NaN         NaN   7.345260         NaN  
...               ...   ...         ...        ...         ...  
2024-05-14  18.097662  8.56  246.880005  55.560001  176.820007  
2024-05-15  17.643988  8.20  251.309998  55.790001  181.130005  
2024-05-16  18.087799  8.23  256.570007  55.849998  179.309998  
2024-05-17  17.930000  8.05  257.929993  55.520000  178.860001  
2024-05-20  17.820000  8.09  259.500000  56.070000  180.600006  

[6134 rows x 101 columns]

---
<font color=green>Q4: (3 Marks)</font>
<br><font color='green'>
Write a Python script to analyze stock data stored in a dictionary `stock_data` (where each key is a stock ticker symbol, and each value is a Pandas Series of adjusted closing prices). The script should:
1. Convert the dictionary into a DataFrame.
2. Calculate the daily returns for each stock.
3. Identify columns (ticker symbols) with at least 2000 non-NaN values in their daily returns.
4. Create a new DataFrame that only includes these filtered ticker symbols.
5. Remove any remaining rows with NaN values in this new DataFrame.
</font>

---

In [26]:
def analyze_stock_data(stock_data):

    stock_df = pd.DataFrame(stock_data)
    daily_returns = stock_df.pct_change()
    valid_columns = daily_returns.columns[daily_returns.count() >= 2000]
    filtered_df = daily_returns[valid_columns]
    cleaned_df = filtered_df.dropna()

    return cleaned_df

stock_data = df.to_dict(orient='series')
cleaned_data = analyze_stock_data(stock_data)
cleaned_data

ADBE       ADP     GOOGL      GOOG      AMZN       AMD  \
Date                                                                     
2015-12-10 -0.006699  0.006004 -0.003292 -0.002861 -0.003715  0.042553   
2015-12-11  0.027653 -0.024924 -0.012657 -0.014130 -0.033473 -0.036735   
2015-12-14  0.020127  0.015240  0.016151  0.012045  0.027744 -0.008475   
2015-12-15  0.008149  0.010284 -0.003213 -0.005844  0.001110  0.008547   
2015-12-16  0.016380  0.008541  0.021708  0.019761  0.026008  0.076271   
...              ...       ...       ...       ...       ...       ...   
2024-05-14 -0.014821 -0.009282  0.007095  0.006027  0.002680  0.017269   
2024-05-15  0.019750  0.004562  0.012739  0.011342 -0.005773  0.042505   
2024-05-16 -0.005089  0.013949  0.009681  0.008914 -0.012689  0.018476   
2024-05-17  0.001139  0.009078  0.010793  0.010603  0.005827  0.011376   
2024-05-20  0.002606 -0.002180  0.004885  0.006599 -0.006280  0.011309   

                 AEP      AMGN       ADI      ANSS  ...      TTWO      TMUS  \
Date                                                ...                       
2015-12-10 -0.024355  0.011274  0.008826  0.004968  ... -0.004777  0.007485   
2015-12-11 -0.005831 -0.028308 -0.003849 -0.013951  ... -0.011575 -0.009356   
2015-12-14 -0.000367  0.019078 -0.001932  0.003899  ...  0.005141  0.014444   
2015-12-15  0.027503  0.028525 -0.004752  0.009544  ...  0.023018  0.044907   
2015-12-16  0.017309  0.012053  0.017330  0.008354  ...  0.006389  0.025419   
...              ...       ...       ...       ...  ...       ...       ...   
2024-05-14 -0.007976  0.009596  0.017084 -0.007130  ...  0.007016 -0.005755   
2024-05-15  0.012997  0.024502  0.017977  0.012337  ...  0.021523  0.001662   
2024-05-16  0.006198 -0.006534 -0.007555 -0.007124  ... -0.013506  0.005532   
2024-05-17  0.001405 -0.007149 -0.000187  0.000550  ...  0.012048  0.002568   
2024-05-20 -0.000863  0.006625  0.015882  0.002685  ...  0.020698 -0.000610   

                TSLA       TXN      VRSK      VRTX       WBA       WBD  \
Date                                                                     
2015-12-10  0.011358  0.002995 -0.002616  0.010279  0.000960 -0.002109   
2015-12-11 -0.044259 -0.012647 -0.015995 -0.034709 -0.020499 -0.035928   
2015-12-14  0.007188  0.000178  0.014390 -0.016491  0.010403 -0.033613   
2015-12-15  0.011483  0.023657  0.013923  0.013656 -0.005450  0.002646   
2015-12-16  0.060699  0.009036  0.021116  0.010658  0.031664  0.024887   
...              ...       ...       ...       ...       ...       ...   
2024-05-14  0.032928  0.017623  0.002395 -0.003117  0.012693  0.021480   
2024-05-15 -0.020051  0.023021  0.003685  0.020766 -0.025068 -0.042056   
2024-05-16  0.004885 -0.002864  0.014687  0.007200  0.025154  0.003659   
2024-05-17  0.014985  0.000256  0.000557  0.010371 -0.008724 -0.021871   
2024-05-20 -0.014144  0.021434  0.003140  0.001482 -0.006135  0.004969   

                WDAY       XEL  
Date                            
2015-12-10  0.005037 -0.013889  
2015-12-11 -0.057630  0.004311  
2015-12-14  0.000253  0.010017  
2015-12-15 -0.000380  0.007935  
2015-12-16  0.029378  0.023896  
...              ...       ...  
2024-05-14 -0.000809 -0.004836  
2024-05-15  0.017944  0.004140  
2024-05-16  0.020930  0.001075  
2024-05-17  0.005301 -0.005909  
2024-05-20  0.006087  0.009906  

[2124 rows x 89 columns]

---
<font color=green>Q5: (1 Mark)</font>
<br><font color='green'>
Download the dataset named `df_filtered_nasdaq_100` from the GitHub repository of the course.
</font>

---

In [27]:
df = pd.read_csv('df_filtered_nasdaq_100.csv')

Date      ADBE       ADP     GOOGL      GOOG      AMZN       AMD  \
0     2015-12-10 -0.006699  0.006004 -0.003292 -0.002861 -0.003715  0.042553   
1     2015-12-11  0.027653 -0.024924 -0.012657 -0.014130 -0.033473 -0.036735   
2     2015-12-14  0.020127  0.015241  0.016151  0.012045  0.027744 -0.008475   
3     2015-12-15  0.008149  0.010283 -0.003213 -0.005844  0.001110  0.008547   
4     2015-12-16  0.016380  0.008541  0.021708  0.019761  0.026008  0.076271   
...          ...       ...       ...       ...       ...       ...       ...   
2113  2024-05-06  0.015241  0.003514  0.005142  0.004971  0.013372  0.034396   
2114  2024-05-07 -0.002674  0.009805  0.018739  0.018548  0.000318 -0.008666   
2115  2024-05-08 -0.008471 -0.008894 -0.010920 -0.010521 -0.004026 -0.005245   
2116  2024-05-09 -0.011166  0.009097  0.003424  0.002454  0.007979 -0.008007   
2117  2024-05-10 -0.000746  0.006975 -0.007708 -0.007518 -0.010660 -0.003084   

           AEP      AMGN       ADI  ...      TTWO      TMUS      TSLA  \
0    -0.024356  0.011274  0.008826  ... -0.004777  0.007485  0.011358   
1    -0.005831 -0.028308 -0.003850  ... -0.011575 -0.009356 -0.044259   
2    -0.000366  0.019078 -0.001932  ...  0.005141  0.014444  0.007188   
3     0.027503  0.028525 -0.004752  ...  0.023018  0.044907  0.011483   
4     0.017309  0.012053  0.017330  ...  0.006389  0.025419  0.060699   
...        ...       ...       ...  ...       ...       ...       ...   
2113  0.002370 -0.037939  0.018484  ...  0.016863 -0.013548  0.019703   
2114  0.011936  0.002738  0.001230  ... -0.000067 -0.001109 -0.037616   
2115  0.007900  0.023343  0.006337  ... -0.015910  0.003946 -0.017378   
2116  0.004085  0.018060 -0.000342  ... -0.001987  0.011361 -0.015739   
2117  0.007257 -0.008662  0.011719  ...  0.001373 -0.002915 -0.020352   

           TXN      VRSK      VRTX       WBA       WBD      WDAY       XEL  
0     0.002996 -0.002616  0.010279  0.000960 -0.002109  0.005037 -0.013889  
1    -0.012648 -0.015996 -0.034709 -0.020499 -0.035928 -0.057630  0.004311  
2     0.000178  0.014390 -0.016491  0.010402 -0.033613  0.000253  0.010017  
3     0.023657  0.013923  0.013656 -0.005450  0.002646 -0.000380  0.007934  
4     0.009036  0.021117  0.010658  0.031664  0.024887  0.029378  0.023897  
...        ...       ...       ...       ...       ...       ...       ...  
2113  0.015427  0.019087  0.003540 -0.030881 -0.001255 -0.022949  0.002028  
2114  0.012752  0.021252  0.019230  0.005214 -0.023869 -0.001921  0.012141  
2115  0.007007 -0.009838  0.020915 -0.006916  0.003861  0.000802 -0.001636  
2116  0.007448  0.001676  0.000406  0.001161  0.030769 -0.014702  0.005644  
2117  0.009335  0.013593  0.009046 -0.003478  0.013682  0.001545  0.003983  

[2118 rows x 90 columns]

---
<font color=green>Q6: (3 Marks) </font>
<br><font color='green'>
Conduct an in-depth analysis of the `df_filtered_nasdaq_100` dataset from GitHub. Answer the following questions:
- Which stock had the best performance over the entire period?
- What is the average daily return of 'AAPL'?
- What is the worst daily return? Provide the stock name and the date it occurred.
</font>

---

In [41]:
# 1. Which stock had the best performance over the entire period?
# Calculate the total return for each stock
total_returns = (df.iloc[-1] / df.iloc[0]) - 1
best_performance_stock = total_returns.idxmax()
best_performance_value = total_returns.max()
print("Stock with the best performance over the entire period:", best_performance_stock, ", Best performance value:", best_performance_value)

# 2. What is the average daily return of 'AAPL'?
daily_returns = df.pct_change()
average_daily_return_aapl = daily_returns['AAPL'].mean()
print("Average daily return of AAPL:", average_daily_return_aapl)

# 3. What is the worst daily return? Provide the stock name and the date it occurred.
worst_return = daily_returns.min().min()
worst_stock = daily_returns.min().idxmin()
worst_date = daily_returns[daily_returns == worst_return].stack().index.tolist()[0][0]
print(f"The worst daily return was {worst_return} by {worst_stock} on {worst_date}")

Stock with the best performance over the entire period: MNST , Best performance value: 1191.734069081319
Average daily return of AAPL: 0.0011996403192543487
The worst daily return was -0.5655690229383197 by REGN on 2003-03-31 00:00:00


# Fama French Analysis

The Fama-French five-factor model is an extension of the classic three-factor model used in finance to describe stock returns. It is designed to better capture the risk associated with stocks and explain differences in returns. This model includes the following factors:

1. **Market Risk (MKT)**: The excess return of the market over the risk-free rate. It captures the overall market's premium.
2. **Size (SMB, "Small Minus Big")**: The performance of small-cap stocks relative to large-cap stocks.
3. **Value (HML, "High Minus Low")**: The performance of stocks with high book-to-market values relative to those with low book-to-market values.
4. **Profitability (RMW, "Robust Minus Weak")**: The difference in returns between companies with robust (high) and weak (low) profitability.
5. **Investment (CMA, "Conservative Minus Aggressive")**: The difference in returns between companies that invest conservatively and those that invest aggressively.

## Additional Factor

6. **Momentum (MOM)**: This factor represents the tendency of stocks that have performed well in the past to continue performing well, and the reverse for stocks that have performed poorly.

### Mathematical Representation

The return of a stock $R_i^t$ at time $t$ can be modeled as follows :

$$
R_i^t - R_f^t = \alpha_i^t + \beta_{i,MKT}^t(R_M^t - R_f^t) + \beta_{i,SMB}^t \cdot SMB^t + \beta_{i,HML}^t \cdot HML^t + \beta_{i,RMW}^t \cdot RMW^t + \beta_{i,CMA}^t \cdot CMA^t + \beta_{i,MOM}^t \cdot MOM^t + \epsilon_i^t
$$

Where:
- $ R_i^t $ is the return of stock $i$ at time $t$
- $R_f^t $is the risk-free rate at time $t$
- $ R_M^t $ is the market return at time $t$
- $\alpha_i^t $ is the abnormal return or alpha of stock $ i $ at time $t$
- $\beta^t $ coefficients represent the sensitivity of the stock returns to each factor at time $t$
- $\epsilon_i^t $ is the error term or idiosyncratic risk unique to stock $ i $ at time $t$

This model is particularly useful for identifying which factors significantly impact stock returns and for constructing a diversified portfolio that is optimized for given risk preferences.




---
<font color=green>Q7: (1 Mark) </font>
<br><font color='green'>
Download the `fama_french_dataset` from the course's GitHub account.
</font>

---

In [32]:
fama = pd.read_csv("fama_french_dataset.csv")

---
<font color=green>Q8: (5 Marks)</font>
<br><font color='green'>

Write a Python function called `get_sub_df_ticker(ticker, date, df_filtered, length_history)` that extracts a historical sub-dataframe for a given `ticker` from `df_filtered`. The function should use `length_history` to determine the number of trading days to include, ending at the specified `date`. Return the sub-dataframe for the specified `ticker`.
</font>

---


---
<font color=green>Q9: (4 Marks)</font>
<br><font color='green'>
Create a Python function named `df_ticker_with_fama_french(ticker, date, df_filtered, length_history, fama_french_data)` that uses `get_sub_df_ticker` to extract historical data for a specific `ticker`. Incorporate the Fama-French factors from `fama_french_data` into the extracted sub-dataframe. Adjust the ticker's returns by subtracting the risk-free rate ('RF') and add other relevant Fama-French factors ('Mkt-RF', 'SMB', 'HML', 'RMW', 'CMA', and 'Mom'). Return the resulting sub-dataframe.
</font>

---

In [ ]:
## Insert your code here

---
<font color=green>Q10: (5 Marks) </font>
<br><font color='green'>
Write a Python function named `extract_beta_fama_french` to perform a rolling regression analysis for a given stock at a specific time point using the Fama-French model. The function should accept the following parameters:

- `ticker`: A string indicating the stock symbol.
- `date`: A string specifying the date for the analysis.
- `length_history`: An integer representing the number of days of historical data to include.
- `df_filtered`: A pandas DataFrame (assumed to be derived from question 5) containing filtered stock data.
- `fama_french_data`: A pandas DataFrame (assumed to be from question 7) that includes Fama-French factors.

Utilize the `statsmodels.api` library to conduct the regression.
</font>

---

In [ ]:
## Insert your code here

---
<font color=green>Q11: (2 Marks) </font>
<br><font color='green'>
Apply the `extract_beta_fama_french` function to the stock symbol 'AAPL' for the date '2024-03-28', using a historical data length of 252 days. Ensure that the `df_filtered` and `fama_french_data` DataFrames are correctly prepared and available in your environment before executing this function. The parameters for the function call are set as follows:

- **Ticker**: 'AAPL'
- **Date**: '2024-03-28'
- **Length of History**: 252 days
</font>

---



In [ ]:
## Insert your code here

---
<font color=green>Q12: (2 Marks)</font>
<br><font color='green'>
Once the `extract_beta_fama_french` function has been applied to 'AAPL' with the specified parameters, the next step is to analyze the regression summary to identify which Fama-French factor explains the most variance in 'AAPL' returns during the specified period.

Follow these steps to perform the analysis:

1. **Review the Summary**: Examine the regression output, focusing on the coefficients and their statistical significance (p-values).
2. **Identify Key Factor**: Determine which factor has the highest absolute coefficient value and is statistically significant (typically p < 0.05). This factor can be considered as having the strongest influence on 'AAPL' returns for the period.

</font>

---

**Write your answers here:**

# PCA Analysis


In literature, another method exists for extracting residuals for each stock, utilizing the PCA approach to identify hidden factors in the data. Let's describe this method.

The return of a stock $R_i^t$ at time $t$ can be modeled as follows :

$$
R_i^t  = \sum_{j=1}^m\beta_{i,j}^t F_j^t  + \epsilon_i^t
$$

Where:
- $ R_i^t $ is the return of stock $i$ at time $t$
- $m$ is the number of factors selected from PCA
-  $ F_j^t $ is the $j$-th hidden factor constructed from PCA at time $t$
- $\beta_{i,j}^t $ are the coefficients representing the sensitivity of the stock returns to each hidden factor.
- $\epsilon_i^t $  is the residual term for stock $i$ at time $t$, representing the portion of the return not explained by the PCA factors.

### Representation of Stock Return Data

Consider the return data for $N$ stocks over $T$ periods, represented by the matrix $R$ of size $T \times N$:

$$
R = \left[
\begin{array}{cccc}
R_1^T & R_2^T & \cdots & R_N^T \\
R_1^{T-1} & R_2^{T-1} & \cdots & R_N^{T-1} \\
\vdots & \vdots & \ddots & \vdots \\
R_1^1 & R_2^1 & \cdots & R_N^1 \\
\end{array}
\right]
$$

Each element $R_i^k$ of the matrix represents the return of stock $i$ at time $k$ and is defined as:

$$
R_i^k = \frac{S_{i,k} - S_{i, k-1}}{S_{i, k-1}}, \quad k=1,\cdots, T, \quad i=1,\cdots,N
$$

where $S_{i,k}$ denotes the adjusted close price of stock $i$ at time $k$.

### Standardization of Returns

To adjust for varying volatilities across stocks, we standardize the returns as follows:

$$
Z_i^t = \frac{R_i^t - \mu_i}{\sigma_i}
$$

where $\mu_i$ and $\sigma_i$ are the mean and standard deviation of returns for stock $i$ over the period $[t-T, t]$, respectively.

### Empirical Correlation Matrix

The empirical correlation matrix $C$ is computed from the standardized returns:

$$
C = \frac{1}{T-1} Z^T Z
$$

where $Z^T$ is the transpose of matrix $Z$.

### Singular Value Decomposition (SVD)

We apply Singular Value Decomposition to the correlation matrix $C$:

$$
C = U \Sigma V^T
$$

Here, $U$ and $V$ are orthogonal matrices representing the left and right singular vectors, respectively, and $\Sigma$ is a diagonal matrix containing the singular values, which are the square roots of the eigenvalues.

### Construction of Hidden Factors

For each of the top $m$ components, we construct the selected hidden factors as follows:

$$
F_j^t = \sum_{i=1}^N \frac{\lambda_{i,j}}{\sigma_i} R_i^t
$$

where $\lambda_{i,j}$ is the $i$-th component of the $j$-th eigenvector (ranked by eigenvalue magnitude).


---
<font color=green>Q13 (3 Marks):

For the specified period from March 29, 2023 ('2023-03-29'), to March 28, 2024 ('2024-03-28'), generate the matrix $Z$ by standardizing the stock returns using the DataFrame `df_filtered_new`
</font>

---


In [ ]:
## Insert your code here

---
<font color=green>Q14: (1 Mark) </font>
<br><font color='green'>
Download the `Z_matrix` matrix from the course's GitHub account.
</font>

---

In [ ]:
## Insert your code here

---
<font color=green>Q15: (3 Marks) </font>
<br><font color='green'>
For the specified period from March 29, 2023 ('2023-03-29'), to March 28, 2024 ('2024-03-28'), compute the correlation matrix
$C$ using the matrix `Z_matrix`.
</font>

---

In [ ]:
## Insert your code here

---
<font color=green>Q16: (2 Marks) </font>
<br><font color='green'>
Refind the correlation matrix from the from March 29, 2023 ('2023-03-29'), to March 28, 2024 ('2024-03-28') using pandas correlation matrix method.
</font>

---

In [ ]:
## Insert your code here

---
<font color=green>Q17: (7 Marks) </font>
<br><font color='green'>
Conduct Singular Value Decomposition on the correlation matrix $C$. Follow these steps:


1.   **Perform SVD**: Decompose the matrix $C$ into its singular values and vectors.
2.   **Rank Eigenvalues**: Sort the resulting singular values (often squared to compare to eigenvalues) in descending order.
3. **Select Components**: Extract the first 20 components based on the largest singular values.
4. **Variance Explained**: Print the variance explained by the first 20 Components and dimensions of differents matrix that you created.

</font>

---

In [ ]:
## Insert your code here

---
<font color=green>Q18: (6 Marks) </font>
<br><font color='green'>
Extract the 20 hidden factors in a matrix F. Check that shape of F is $(252,20)$
</font>

</font>

---

In [ ]:
## Insert your code here

---
<font color=green>Q19: (3 Marks) </font>
<br><font color='green'>
Perform the Regression Analysis of 'AAPL' for the date '2024-03-28', using a historical data length of 252 days using previous $F$ Matrix. Compare the R-squared from the ones obtained at Q11.
</font>

</font>

---

In [ ]:
## Insert your code here

# Ornstein Uhlenbeck

The Ornstein-Uhlenbeck process is defined by the following stochastic differential equation (SDE):

$$ dX_t = \theta (\mu - X_t) dt + \sigma dW_t $$

where:

- **$ X_t $**: The value of the process at time $ t $.
- **$ \mu $**: The long-term mean (equilibrium level) to which the process reverts.
- **$ \theta $**: The speed of reversion or the rate at which the process returns to the mean.
- **$ \sigma $**: The volatility (standard deviation), representing the magnitude of random fluctuations.
- **$ W_t $**: A Wiener process or Brownian motion that adds stochastic (random) noise.

This equation describes a process where the variable $ X_t $ moves towards the mean $ \mu $ at a rate determined by $ \theta $, with random noise added by $ \sigma dW_t $.

---
<font color=green>Q20: (3 Marks) </font>
<br><font color='green'>
In the context of mean reversion, which quantity should be modeled using an Ornstein-Uhlenbeck process?
</font>

---

**Write your answers here:**

---
<font color=green>Q21: (5 Marks) </font>
<br><font color='green'>
Explain how the parameters $ \theta $ and $ \sigma $ can be determined using the following equations. Also, detail the underlying assumptions:
$$ E[X] = \mu $$
$$ \text{Var}[X] = \frac{\sigma^2}{2\theta} $$
</font>

---

**Write your answers here:**

---
<font color=green>Q22: (2 Marks) </font>
<br><font color='green'>
Create a function named `extract_s_scores` which computes 's scores' for the last element in a list of floating-point numbers. This function calculates the scores using the following formula $ \text{s scores} = \frac{X_T - \mu}{\sigma} $ where `list_xi` represents a list containing a sequence of floating-point numbers $(X_0, \cdots, X_T)$.

</font>

---

In [ ]:
## Insert your code here

# Autoencoder Analysis

Autoencoders are neural networks used for unsupervised learning, particularly for dimensionality reduction and feature extraction. Training an autoencoder on the $Z_i$ matrix aims to identify hidden factors capturing the intrinsic structures in financial data.

### Architecture
- **Encoder**: Compresses input data into a smaller latent space representation.
  - *Input Layer*: Matches the number of features in the $Z_i$ matrix.
  - *Hidden Layers*: Compress data through progressively smaller layers.
  - *Latent Space*: Encodes the data into hidden factors.
- **Decoder**: Reconstructs input data from the latent space.
  - *Hidden Layers*: Gradually expand to the original dimension.
  - *Output Layer*: Matches the input layer to recreate the original matrix.

### Training
The autoencoder is trained by minimizing reconstruction loss, usually mean squared error (MSE), between the input $Z_i$ matrix and the decoder's output.

### Hidden Factors Extraction
After training, the encoder's latent space provides the most important underlying patterns in the stock returns.

---
<font color=green>Q23: (2 Marks) </font>
<br><font color='green'>
Modify the standardized returns matrix `Z_matrix` to reduce the influence of extreme outliers on model trainingby ensuring that all values in the matrix `Z_matrix` do not exceed 3 standard deviations from the mean. Specifically, cap these values at the interval $-3, 3]$. Store the adjusted values in a new matrix, `Z_hat`.
</font>

----

In [ ]:
## Insert your code here

---
<font color=green>Q24: (1 Marks) </font>
<br><font color='green'>
Fetch the `Z_hat` data from GitHub, and we'll proceed with it now.
</font>



In [ ]:
## Insert your code here

---
<font color=green>Q25: (3 Marks) </font>
<br><font color='green'>
Segment the standardized and capped returns matrix $\hat{Z}$ into two subsets for model training and testing. Precisly Allocate 70% of the data in $\hat{Z}$ to the training set $ \hat{Z}_{train} $ and Allocate the remaining 30% to the testing set $\hat{Z}_{test}$. Treat each stock within $\hat{Z}$ as an individual sample, by flattening temporal dependencies.
</font>



In [ ]:
## Insert your code here

---
<font color=green>Q26: (10 Marks) </font>
<br><font color='green'>
Please create an autoencoder following the instructions provided in  **[End-to-End Policy Learning of a Statistical Arbitrage Autoencoder Architecture](https://arxiv.org/pdf/2402.08233.pdf)**, Use the model 'Variant 2' in Table 1.
</font>

---

In [ ]:
## Insert your code here

---
<font color=green>Q27 (1 Mark) :

Display all the parameters of the deep neural network.
</font>

---

In [ ]:
## Insert your code here

---
<font color=green>Q28: (3 Marks) </font>
<br><font color='green'>
Train your model using the Adam optimizer for 20 epochs with a batch size equal to 8 and validation split to 20%. Specify the loss function you've chosen.
</font>


In [ ]:
## Insert your code here

---
<font color=green>Q29: (3 Marks) </font>
<br><font color='green'>
Predict using the testing set and extract the residuals based on the methodology described in **[End-to-End Policy Learning of a Statistical Arbitrage Autoencoder Architecture](https://arxiv.org/pdf/2402.08233.pdf)**.
for 'NVDA' stock.
</font>

---

In [ ]:
## Insert your code here

<font color=green>Q30: (7 Marks) </font>
<br><font color='green'>
By reading carrefully the paper **[End-to-End Policy Learning of a Statistical Arbitrage Autoencoder Architecture](https://arxiv.org/pdf/2402.08233.pdf)**, answers the following question:
1. **Summarize the Key Actions**: Highlight the main experiments and methodologies employed by the authors in Section 5.
2. **Reproduction Steps**: Detail the necessary steps required to replicate the authors' approach based on the descriptions provided in the paper.
3. **Proposed Improvement**: Suggest one potential enhancement to the methodology that could potentially increase the effectiveness or efficiency of the model.



**Write your answers here:**








